lib/Sema/ConstraintSystem.cpp

```diff
@@ -2045,6 +2045,15 @@ bool ConstraintSystem::salvage(SmallVectorImpl<Solution> &viable, Expr *expr) {
     SolverState state(expr, *this, FreeTypeVariableBinding::Disallow);
     state.recordFixes = true;
 
+    // Retry all inactive and failed constraints
+    if (failedConstraint) {
+      addUnsolvedConstraint(failedConstraint);
+      failedConstraint = nullptr;
+    }
+    ActiveConstraints.splice(ActiveConstraints.end(), InactiveConstraints);
+    for (auto &constraint : ActiveConstraints)
+      constraint.setActive(true);
```

Comment:
I still think that most of this logic is unnecessary because failed constraint should either still be in the set of inactive constraints, or it would be re-created in case it was generated and erased. I'm going to take a deeper look today, to make sure I'm not wrong :)

Reply:
That's a bug then... because https://github.com/apple/swift/blob/master/lib/Sema/CSSolver.cpp#L339L342 constraints should be retired and re-inserted back to the inactive list... It seems like the problem might be that there is no scope so it just nothing is being rolled back in that case.

Reply:
So yes, that's exactly problem here! Before attempting constraints we need to create a scope to return all of the retired constraints back at the end and clear a failed constraint, it would make sense for solver state to record which constraints were active at the time of its creation to re-activate them on destruction. Working on the patch for this right now...